smart-contracts/contracts/cl-vault/src/rewards/distribution.rs

```diff
@@ -87,80 +116,119 @@ pub fn handle_collect_spread_rewards_reply(
+
+    let mut users_processed: u128 = 0;
+
+    while users_processed < amount_of_users.u128() {
+        // Attempt to pop a (user_address, share_amount) tuple from the front of the snapshot Deque
+        if let Some((user_address, share_amount)) = DISTRIBUTION_SNAPSHOT.pop_front(deps.storage)? {
+            debug!(deps, "user_address", user_address);
+            debug!(deps, "share_amount", share_amount);
+
+            // Calculate the user's reward based on the share amount directly
+            let reward_ratio = Decimal::from_ratio(share_amount, total_shares.u128());
+            let user_reward = rewards.mul_ratio(reward_ratio);
+            let user_reward_clone = user_reward.clone();
+
+            debug!(deps, "user_reward_clone", user_reward_clone);
+
+
+            // Merge the new reward with any existing rewards for the user
+            USER_REWARDS.update(
+                deps.storage,
+                user_address,
+                |old| -> ContractResult<CoinList> {
+                    Ok(match old {
+                        Some(old_rewards) => user_reward_clone.add(old_rewards)?,
+                        None => user_reward_clone,
+                    })
+                },
+            )?;
+
+            // Accumulate the distributed rewards to be subtracted later
+            distributed_rewards.merge(user_reward.coins())?;
+            debug!(deps, "distributed_rewards", distributed_rewards);
+            users_processed += 1;
+        } else {
+            // No more addresses in the snapshot Deque to process
+            break;
+        }
+    }
 
-    let bq = BankQuerier::new(&deps.querier);
-    let vault_denom = VAULT_DENOM.load(deps.storage)?;
-
-    let total_shares: Uint128 = bq
-        .supply_of(vault_denom)?
-        .amount
-        .unwrap()
-        .amount
-        .parse::<u128>()?
-        .into();
-
-    // for each user with locked tokens, we distribute some part of the rewards to them
-    // get all users and their current pre-distribution rewards
-    let user_rewards: Result<Vec<(Addr, CoinList)>, ContractError> = SHARES
-        .range(deps.branch().storage, None, None, Order::Ascending)
-        .map(|v| -> Result<(Addr, CoinList), ContractError> {
-            let (address, user_shares) = v?;
-            // calculate the amount of each asset the user should get in rewards
-            // we need to always round down here, so we never expect more rewards than we have
-            let user_rewards = rewards.mul_ratio(Decimal::from_ratio(user_shares, total_shares));
-            Ok((address, user_rewards))
-        })
-        .collect();
+    // If the Deque is empty, we've finished distributing rewards so we clear the semaphore flag
+    let is_last_distribution = DISTRIBUTION_SNAPSHOT.is_empty(deps.storage)?;
+    if is_last_distribution {
+        IS_DISTRIBUTING.save(deps.storage, &false)?;
+        HAS_FEE_BEEN_CALCULATED.save(deps.storage, &false)?;
```

Comment:
I don't see the need for this line. we aren't setting anywhere

Reply:
Yes this HAS_FEE_BEEN_CALCULATED  has been deprecated after removing strategist calculation from the distribution. Or were you referring to another?

Reply:
yes that one